cmd/osbuild-worker/main.go

```diff
@@ -154,6 +184,89 @@ func RunJob(job *worker.Job, store string, uploadFunc func(uuid.UUID, string, io
 				azureMaxUploadGoroutines,
 			)
 
+			if err != nil {
+				r = append(r, err)
+				continue
+			}
+		case *target.KojiTargetOptions:
+			// Koji for some reason needs TLS renegotiation enabled.
+			// Clone the default http transport and enable renegotiation.
+			transport := http.DefaultTransport.(*http.Transport).Clone()
+			transport.TLSClientConfig = &tls.Config{
+				Renegotiation: tls.RenegotiateOnceAsClient,
+			}
+
+			k, err := koji.NewFromGSSAPI(options.Server, &koji.GSSAPICredentials{}, transport)
+			if err != nil {
+				r = append(r, err)
+				continue
+			}
+
+			defer func() {
+				err := k.Logout()
+				if err != nil {
+					log.Printf("koji logout failed: %v", err)
+				}
+			}()
+
+			f, err := os.Open(path.Join(outputDirectory, options.Filename))
+			if err != nil {
+				r = append(r, err)
+				continue
+			}
+
+			hash, filesize, err := k.Upload(f, options.UploadDirectory, options.Filename)
+			if err != nil {
+				r = append(r, err)
+				continue
+			}
+
+			build := koji.ImageBuild{
+				BuildID:   options.BuildID,
+				Name:      options.Name,
+				Version:   options.Version,
+				Release:   options.Release,
+				StartTime: start_time.Unix(),
+				EndTime:   end_time.Unix(),
+			}
+			buildRoots := []koji.BuildRoot{
+				{
+					ID: 1,
+					Host: koji.Host{
+						Os:   "RHEL8",
+						Arch: "noarch",
+					},
+					ContentGenerator: koji.ContentGenerator{
+						Name:    "osbuild",
+						Version: "1",
+					},
+					Container: koji.Container{
+						Type: "nspawn",
+						Arch: "noarch",
+					},
+					Tools: []koji.Tool{},
+					RPMs:  osbuildStagesToRPMs(result.Build.Stages),
+				},
+			}
+			output := []koji.Image{
+				{
+					BuildRootID:  1,
+					Filename:     options.Filename,
+					FileSize:     uint64(filesize),
+					Arch:         "noarch",
+					ChecksumType: "md5",
+					MD5:          hash,
+					Type:         "image",
+					RPMs:         osbuildStagesToRPMs(result.Stages),
+					Extra: koji.ImageExtra{
+						Info: koji.ImageExtraInfo{
+							Arch: "noarch",
+						},
+					},
+				},
+			}
+
+			_, err = k.CGImport(build, buildRoots, output, options.UploadDirectory, options.Token)
 			if err != nil {
 				r = append(r, err)
 				continue
```

Comment:
Need to `CGRefundBuild` here if it goes wrong, i.e. something this happens:

```
Sep 09 12:37:50 snow osbuild-worker[44822]: error: "Traceback (most recent call last):
Sep 09 12:37:50 snow osbuild-worker[44822]: error: "Traceback (most recent call last):
Sep 09 12:37:50 snow osbuild-worker[44822]:   File "/usr/share/koji-hub/kojihub.py", line 6295, in check_build_dir
Sep 09 12:37:50 snow osbuild-worker[44822]:     raise koji.GenericError("Destination directory already exists: %s" % path)
Sep 09 12:37:50 snow osbuild-worker[44822]: koji.GenericError: Destination directory already exists: /mnt/koji/packages/Fedora-Cloud/32/1
Sep 09 12:37:50 snow osbuild-worker[44822]: Locals by frame, innermost last
Sep 09 12:37:50 snow osbuild-worker[44822]: Frame application in /usr/share/koji-hub/kojixmlrpc.py at line 752
Sep 09 12:37:50 snow osbuild-worker[44822]:                environ = {'UNIQUE_ID': 'X1iwfS59HN9FZPa4qsQ7UwAAANU', 'SSL_TLS_SNI': 'localhost', 'GATEWAY_INTERFACE': 'CGI/1.1', 'SERVER_PROTOCOL': 'HTTP/1>
Sep 09 12:37:50 snow osbuild-worker[44822]:         start_response = <built-in method start_response of mod_wsgi.Adapter object at 0x7f01d1404f30>
Sep 09 12:37:50 snow osbuild-worker[44822]:                  start = 1599647869.4933336
Sep 09 12:37:50 snow osbuild-worker[44822]:   memory_usage_at_start = 31824
Sep 09 12:37:50 snow osbuild-worker[44822]:                      h = <_mod_wsgi_df1c13776110deb1e020020636703560.ModXMLRPCRequestHandler object at 0x7f01d0dcfdc0>
Sep 09 12:37:50 snow osbuild-worker[44822]: Frame _wrap_handler in /usr/share/koji-hub/kojixmlrpc.py at line 263
Sep 09 12:37:50 snow osbuild-worker[44822]:                   self = <_mod_wsgi_df1c13776110deb1e020020636703560.ModXMLRPCRequestHandler object at 0x7f01d0dcfdc0>
Sep 09 12:37:50 snow osbuild-worker[44822]:                handler = <bound method ModXMLRPCRequestHandler.handle_rpc of <_mod_wsgi_df1c13776110deb1e020020636703560.ModXMLRPCRequestHandler object at 0>
Sep 09 12:37:50 snow osbuild-worker[44822]:                environ = {'UNIQUE_ID': 'X1iwfS59HN9FZPa4qsQ7UwAAANU', 'SSL_TLS_SNI': 'localhost', 'GATEWAY_INTERFACE': 'CGI/1.1', 'SERVER_PROTOCOL': 'HTTP/1>
Sep 09 12:37:50 snow osbuild-worker[44822]: Frame handle_rpc in /usr/share/koji-hub/kojixmlrpc.py at line 277
Sep 09 12:37:50 snow osbuild-worker[44822]:                   self = <_mod_wsgi_df1c13776110deb1e020020636703560.ModXMLRPCRequestHandler object at 0x7f01d0dcfdc0>
Sep 09 12:37:50 snow osbuild-worker[44822]:                environ = {'UNIQUE_ID': 'X1iwfS59HN9FZPa4qsQ7UwAAANU', 'SSL_TLS_SNI': 'localhost', 'GATEWAY_INTERFACE': 'CGI/1.1', 'SERVER_PROTOCOL': 'HTTP/1>
Sep 09 12:37:50 snow osbuild-worker[44822]:                 params = <ERROR WHILE PRINTING VALUE>
Sep 09 12:37:50 snow osbuild-worker[44822]:                 method = CGImport
Sep 09 12:37:50 snow osbuild-worker[44822]: Frame _dispatch in /usr/share/koji-hub/kojixmlrpc.py at line 314
Sep 09 12:37:50 snow osbuild-worker[44822]:                   self = <_mod_wsgi_df1c13776110deb1e020020636703560.ModXMLRPCRequestHandler object at 0x7f01d0dcfdc0>
Sep 09 12:37:50 snow osbuild-worker[44822]:                 method = CGImport
Sep 09 12:37:50 snow osbuild-worker[44822]:                 params = <ERROR WHILE PRINTING VALUE>
Sep 09 12:37:50 snow osbuild-worker[44822]:                   func = <function cg_import at 0x7f01d162a430>
Sep 09 12:37:50 snow osbuild-worker[44822]:                   opts = {}
Sep 09 12:37:50 snow osbuild-worker[44822]: Frame call_with_argcheck in /usr/lib/python3.8/site-packages/koji/util.py at line 277
Sep 09 12:37:50 snow osbuild-worker[44822]:                   func = <function cg_import at 0x7f01d162a430>
Sep 09 12:37:50 snow osbuild-worker[44822]:                   args = <ERROR WHILE PRINTING VALUE>
Sep 09 12:37:50 snow osbuild-worker[44822]:                 kwargs = {}
Sep 09 12:37:50 snow osbuild-worker[44822]: Frame cg_import in /usr/share/koji-hub/kojihub.py at line 6171
Sep 09 12:37:50 snow osbuild-worker[44822]:               metadata = {"metadata_version":0,"build":{"build_id":1,"name":"Fedora-Cloud","version":"32","release":"1","source":"","start_time":1599647761,>
Sep 09 12:37:50 snow osbuild-worker[44822]:              directory = osbuild-composer-koji-96d6ba59-b5f4-4a4a-9371-6382a3a27727
Sep 09 12:37:50 snow osbuild-worker[44822]:                  token = e941710f8eb363161f11e1819a2c95639cf62e0539d74a8b10019d1dd1f75f99
Sep 09 12:37:50 snow osbuild-worker[44822]:               importer = <kojihub.CG_Importer object at 0x7f01d0dcfcd0>
Sep 09 12:37:50 snow osbuild-worker[44822]: Frame do_import in /usr/share/koji-hub/kojihub.py at line 6199
Sep 09 12:37:50 snow osbuild-worker[44822]:                   self = <kojihub.CG_Importer object at 0x7f01d0dcfcd0>
Sep 09 12:37:50 snow osbuild-worker[44822]:               metadata = {'metadata_version': 0, 'build': {'build_id': 1, 'name': 'Fedora-Cloud', 'version': '32', 'release': '1', 'source': '', 'start_time>
Sep 09 12:37:50 snow osbuild-worker[44822]:              directory = osbuild-composer-koji-96d6ba59-b5f4-4a4a-9371-6382a3a27727
Sep 09 12:37:50 snow osbuild-worker[44822]:                  token = e941710f8eb363161f11e1819a2c95639cf62e0539d74a8b10019d1dd1f75f99
Sep 09 12:37:50 snow osbuild-worker[44822]:                metaver = 0
Sep 09 12:37:50 snow osbuild-worker[44822]: Frame check_build_dir in /usr/share/koji-hub/kojihub.py at line 6295
Sep 09 12:37:50 snow osbuild-worker[44822]:                   self = <kojihub.CG_Importer object at 0x7f01d0dcfcd0>
Sep 09 12:37:50 snow osbuild-worker[44822]:                 delete = False
Sep 09 12:37:50 snow osbuild-worker[44822]:                   path = /mnt/koji/packages/Fedora-Cloud/32/1
```
Because otherwise the build will be stuck at `BUILDING`:
```
BUILD: Fedora-Cloud-32-1 [1]
State: BUILDING
Reserved by: osbuild
Built by: osbuild-krb
Volume: DEFAULT
Task: none
Finished: 
Tags: 
```

Reply:
Fixed.